[JIT] Add comments for adding shape function and linting

eellison opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* __->__ #73570

cc @silvasean 

add lints that the types of shape functions are as expected, and for few other patterns

Differential Revision: [D35192688](https://our.internmc.facebook.com/intern/diff/D35192688)

Commit:
Add comments for adding shape function and linting

[ghstack-poisoned]

Commit:
Update on "[JIT] Add comments for adding shape function and linting"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add comments for adding shape function and linting"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add comments for adding shape function and linting"




[ghstack-poisoned]

Commit:
Update on "[JIT] Add comments for adding shape function and linting"


cc silvasean 

[ghstack-poisoned]

silvasean commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -160,26 +163,130 @@ const at::optional<const FunctionSchema*> getInplaceVariant(
+void checkInputReturnedAsOutput(
+    const FunctionSchema* schema,
+    const std::shared_ptr<Graph>& graph) {
+  // Could use alias db here as well but would have to warn because it's
```
Nice. I think a few of my handwritten shape functions are breaking this invariant! Will try to upstream them and get this great linting :)

eellison replied:
The main thing I'm trying to preserve is that you can compose shape functions together... Technically it's fine if an aliased list is return so long as no shape function mutates its' input, however I thought it would be a little more defensive/safe to return a new unaliased list..

huiguoo commented on torch/csrc/jit/runtime/symbolic_shape_registry.h:
```diff
@@ -8,6 +8,44 @@
+in `test_symbolic_shape_analysis.py` such as `test_adaptive_avg_pool2d`.
+
+Operators which take in a list of tensors, such as concat, are not yet
+supported. Concat has been special cased and could be generalized as needed.
```
I am adding support for `aten::stack` which takes in a list of tensors. Added you as the reviewer for PRs in that stream.

huiguoo commented on torch/csrc/jit/runtime/symbolic_shape_registry.cpp:
```diff
@@ -197,9 +304,31 @@ void registerSchema(
+      toGraphFunction(shape_compute_function).graph();
+
+  transformShapeFunction(schema_string, graph);
+  checkShapeFunction(schema_string, graph);
```
Are all shape compute graphs hard coded? If so, can we move the checks to unit tests instead of doing it every time we load the module? Or perhaps I understand it incorrectly.

huiguoo reviewed (approved):
I have a small question in the comments. Overall, the checks and comments look good to me!

Commit:
Update on "[JIT] Add comments for adding shape function and linting"


cc silvasean 

add lints that the types of shape functions are as expected, and for few other patterns

[ghstack-poisoned]

Commit:
Update on "[JIT] Add comments for adding shape function and linting"


cc silvasean 

add lints that the types of shape functions are as expected, and for few other patterns

[ghstack-poisoned]

Commit:
Update on "[JIT] Add comments for adding shape function and linting"


cc silvasean 

add lints that the types of shape functions are as expected, and for few other patterns

[ghstack-poisoned]

eellison commented:
@eellison has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D35192688).

eellison commented:
@pytorchbot merge this please

jjsjann123 commented on test/jit/test_symbolic_shape_analysis.py:
```diff
@@ -498,3 +499,37 @@ def test_shape_function_includes(self):
         res = torch.jit._shapes.matmul(m1_shape, m2_shape)
         self.assertEqual(res, [10, 10])
+
+    def test_register_function_error_checking(self):
```
Nice python exposure. We actually get to patch it hot on our scripts!

eellison commented:
@pytorchmergebot merge this please

Commit:
Update on "[JIT] Add comments for adding shape function and linting"


cc silvasean 

add lints that the types of shape functions are as expected, and for few other patterns

Differential Revision: [D35192688](https://our.internmc.facebook.com/intern/diff/D35192688)

[ghstack-poisoned]

eellison commented:
@pytorchmergebot merge this please
